### PR TITLE
implemented api

### PR DESCRIPTION
**Pull request by Ragnar-Doorlock**

### CODE REVIEW

File: app/dateProvider/dateProvider.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        }
+
+        return yyyy + '-' + mm + '-' + dd;
+    }
+
+    convertDate(date) {
+        const convertedDate = new Date(date);
+        return convertedDate;
+    }
+}
```
Comment (Manstrike) [inline app/dateProvider/dateProvider.js]: No need to make a bicycle. Use a luxon lib to process dates

File: app/exchangeRateProvider.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+const latestCurrencyUrl = process.env.EXCHANGE_RATES_URL_LATEST;
+const historicalCurrencyUrl = process.env.EXCHANGE_RATES_HISTORICAL;
+const appId = process.env.EXCHANGE_RATES_APP_ID;
```
Comment (Manstrike) [inline app/exchangeRateProvider.js]: These should be provide through constructor

File: app/exchangeRateProvider.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    async getHistoricalRates(date) {
+        const result = await this.axios.get(`${historicalCurrencyUrl}${date}.json?app_id=${appId}`);
+        return result.data.rates;
+    }
+
+    async calculateExchangeRate(rates, currency) {
+        try {
+            this.fx.base = 'USD';
+            this.fx.rates = rates;
+            return await this.fx.convert(1, { from: currency, to: 'UAH' });
```
Comment (Manstrike) [inline app/exchangeRateProvider.js]: `return await...`

Comment (Manstrike) [reply app/exchangeRateProvider.js]: Currencies should be listed in a enum or some mapper.

File: app/exchangeRateRepository.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        if (isRateCached) {
+            const rates = await this.cache.get(date);
+            const parsedRates = JSON.parse(rates);
+            return this.exchangerFactory.create({
+                USDRate: parsedRates.usd,
+                EURRate: parsedRates.eur,
+                date: date
+            });
+        }
+        const result = await this.dbProvider.getData({ date: date });
```
Comment (Manstrike) [inline app/exchangeRateRepository.js]: const [data] = await this....

File: app/get-exchange-rate/getExchangeRateInteractor.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            this.presenter.presentFailure(new ValidationError(errors));
+            return;
+        }
+
+        const rates = await this.exchangeRateRepository.getRateByDate(request.date);
+        if (!rates) {
+            // probably currency api would work if i request actual rates as historical with current date
+            // but let it be that way for now
+            let rates;
+            if (this.dateProvider.getCurrentDate() === request.date) {
```
Comment (Manstrike) [inline app/get-exchange-rate/getExchangeRateInteractor.js]: If the only purpose of DateProvider to convert date to some format, that should be later be compared... than u can use defualt JS Date object)
```suggestion
            if (new Date() === new Date(request.date)) {
```

File: app/get-exchange-rate/getExchangeRateInteractor.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+                date: request.date,
+                usd: resultUSD,
+                eur: resultEUR,
+                rates
+            });
+            this.presenter.presentSuccess(this.responseBuilder.build( entity ));
+            return;
+        }
+
+        const entity = this.exchangerFactory.create({
```
Comment (Manstrike) [inline app/get-exchange-rate/getExchangeRateInteractor.js:57]: Dont u think it is akward to make entity to destroy?) No need to do it if only response builder does not rely on it strictly.
Btw, I think early-returtn pattern can be applied here

File: app/get-exchange-rate/getExchangeRateInteractor.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        loggerProvider,
+        exchangeRateProvider,
+        dateProvider
+    }) {
+        this.presenter = presenter;
+        this.validator = validator;
+        this.exchangerFactory = exchangerFactory;
+        this.exchangeRateRepository = exchangeRateRepository;
+        this.responseBuilder = getExchangeRateResponseBuilder;
+        this.logger = loggerProvider.create(GetExchangeRateInteractor.name);
```
Comment (Manstrike) [inline app/get-exchange-rate/getExchangeRateInteractor.js]: Cannot see where it is used

File: db/dbProvider.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
+class DBProvider {
+    constructor({ client }) {
+        this.client = client;
+    }
+
+    async insert(data) {
+        const db = this.client.db(process.env.MONGO_DB);
```
Comment (Manstrike) [inline db/dbProvider.js]: Pass envs through constructor

Comment (Manstrike) [reply db/dbProvider.js]: btw u can create DBProvider with already selected collection
Then u can create provider duplicates when needed to use another collection

File: app/exchangeRateRepository.js (hunk 4 above)
Comment (Ragnar-Doorlock) [reply app/exchangeRateRepository.js]: why?

**Pull request merged**